File: backend/app/schemas/assessment.py
from pydantic import BaseModel, Field
from typing import Dict, List, Optional


class StartAssessmentRequest(BaseModel):
    skill: str = Field(..., min_length=1, max_length=120, description="Canonical skill name or alias, e.g. 'Python'")
    question_count: Optional[int] = Field(
        None, ge=3, le=7, description="Requested number of questions (prototype range 3-7)"
    )


class SubmitAssessmentRequest(BaseModel):
    attempt_id: str = Field(..., min_length=8, max_length=64)
    responses: Dict[str, str] = Field(
        default_factory=dict,
        description="Map of question_id -> selected option id (MCQ) or short answer text",
    )
    duration_seconds: Optional[int] = Field(None, ge=0, le=86400)


class AssessmentOption(BaseModel):
    id: str
    text: str


class AssessmentQuestionOut(BaseModel):
    id: str
    skill: str
    difficulty: str
    kind: str
    prompt: str
    type: str
    options: List[AssessmentOption] = []
    version: str


class StartAssessmentResponse(BaseModel):
    attempt_id: str
    skill: str
    skill_key: str
    assessment_version: str
    question_count: int
    started_at: str
    expires_at: str
    questions: List[AssessmentQuestionOut]
    disclaimer: str


class LastAssessment(BaseModel):
    attempt_id: Optional[str] = None
    score: float = 0.0
    correct_count: int = 0
    question_count: int = 0
    completed_at: Optional[str] = None
    assessment_version: Optional[str] = None
    total_attempts: int = 1


class LayerCapability(BaseModel):
    """Which assessment layers apply to one skill (additive, optional)."""

    skill: str = ""
    skill_key: str = ""
    knowledge_assessment: bool = False
    practical_assessment: bool = False
    interview_assessment: bool = False
    practical_kind: Optional[str] = None
    layers: List[str] = []


class LayerResult(BaseModel):
    """Completion state of one layer for one skill (additive, optional)."""

    status: str = "not_started"
    score: Optional[float] = None
    attempt_id: Optional[str] = None
    session_id: Optional[str] = None
    completed_at: Optional[str] = None
    validity: Optional[str] = None
    communication_score: Optional[float] = None


class AvailableAssessment(BaseModel):
    skill: str
    skill_key: str
    reason: str
    reason_label: str
    proficiency: float = 0.0
    confidence: float = 0.0
    required_level: float = 0.0
    gap: float = 0.0
    importance: float = 0.0
    priority_score: float = 0.0
    evidence_count: int = 0
    already_assessed: bool = False
    question_count: int = 0
    last_assessment: Optional[LastAssessment] = None
    # 3-layer assessment (additive): applicable layers + per-layer progress.
    capabilities: Optional[LayerCapability] = None
    layers: Optional[Dict[str, LayerResult]] = None


class AvailableAssessmentsResponse(BaseModel):
    assessment_version: str
    target_role: Optional[str] = None
    analysis_id: Optional[str] = None
    available: List[AvailableAssessment] = []
    disclaimer: str


class QuestionFeedback(BaseModel):
    question_id: str
    skill: str
    difficulty: str
    kind: str
    correct: bool
    answered: bool
    explanation: str


class SubmitAssessmentResponse(BaseModel):
    attempt_id: str
    skill: str
    assessment_version: str
    score: float
    correct_count: int
    question_count: int
    validity: str
    counts_as_evidence: bool
    completed_at: str
    source_reliability: float
    feedback: List[QuestionFeedback] = []
    analysis: Optional[dict] = None
    disclaimer: str


# ---------------------------------------------------------------------------
# Layer 2 — practical assessment
# ---------------------------------------------------------------------------

class PracticalTaskDimension(BaseModel):
    id: str
    label: str


class PracticalTaskOut(BaseModel):
    id: str
    skill: str
    task_type: str
    title: str
    prompt: str
    starter_code: str = ""
    language: str = ""
    estimated_minutes: int = 20
    dimensions: List[PracticalTaskDimension] = []
    check_count: int = 0
    version: str


class StartPracticalRequest(BaseModel):
    skill: str = Field(..., min_length=1, max_length=120)


class StartPracticalResponse(BaseModel):
    attempt_id: str
    skill: str
    skill_key: str
    task_version: str
    started_at: str
    task: PracticalTaskOut
    disclaimer: str


class PracticalCheckResult(BaseModel):
    check_id: str
    dimension: str
    description: str
    passed: bool


class SubmitPracticalRequest(BaseModel):
    attempt_id: str = Field(..., min_length=8, max_length=64)
    code: str = Field(default="", max_length=20000)
    duration_seconds: Optional[int] = Field(None, ge=0, le=86400)


class SubmitPracticalResponse(BaseModel):
    attempt_id: str
    skill: str
    task_id: str
    task_version: str
    score: float
    validity: str
    counts_as_evidence: bool
    completed_at: str
    source_reliability: float
    dimensions: Dict[str, float] = {}
    checks: List[PracticalCheckResult] = []
    analysis: Optional[dict] = None
    disclaimer: str


# ---------------------------------------------------------------------------
# Layer 3 — AI skill interview
# ---------------------------------------------------------------------------

class StartInterviewRequest(BaseModel):
    skill: str = Field(..., min_length=1, max_length=120)


class InterviewTTSRequest(BaseModel):
    text: str = Field(..., min_length=1, max_length=4000)
    session_id: str = Field(default="", max_length=64)
    question_id: str = Field(default="", max_length=120)


class InterviewQuestionOut(BaseModel):
    id: str
    competency: str
    prompt: str
    follow_ups: List[str] = []


class InterviewCompetencyOut(BaseModel):
    id: str
    label: str


class StartInterviewResponse(BaseModel):
    session_id: str
    skill: str
    skill_key: str
    interview_version: str
    status: str
    started_at: str
    plan: dict
    evaluated_dimensions: Dict[str, List[str]] = {}
    privacy_notice: str
    disclaimer: str


class InterviewSessionOut(BaseModel):
    session_id: Optional[str] = None
    skill: Optional[str] = None
    status: str = "in_progress"
    plan: dict = {}
    transcript: List[dict] = []
    technical_scores: Optional[dict] = None
    communication_scores: Optional[dict] = None
    started_at: Optional[str] = None
    completed_at: Optional[str] = None


class SubmitInterviewResponsesRequest(BaseModel):
    session_id: str = Field(..., min_length=8, max_length=64)
    responses: Dict[str, str] = Field(
        default_factory=dict,
        description="Map of question_id -> written answer text",
    )


class SubmitInterviewResponsesResponse(BaseModel):
    session_id: str
    skill: Optional[str] = None
    answered: int = 0
    total: int = 0


class CompleteInterviewRequest(BaseModel):
    session_id: str = Field(..., min_length=8, max_length=64)


class CompleteInterviewResponse(BaseModel):
    session_id: str
    skill: str
    status: str
    validity: str
    counts_as_evidence: bool
    completed_at: Optional[str] = None
    source_reliability: Optional[float] = None
    technical_scores: Optional[dict] = None
    communication_scores: Optional[dict] = None
    analysis: Optional[dict] = None
    grading_path: Optional[str] = None
    note: Optional[str] = None


# ---------------------------------------------------------------------------
# Layer 3 — Per-answer adaptive interview
# ---------------------------------------------------------------------------

class AnswerInterviewRequest(BaseModel):
    question_id: str = Field(..., min_length=1, max_length=120)
    transcript: str = Field(..., min_length=1, max_length=8000, description="Answer text (typed or voice-transcribed)")


class AnswerInterviewEvaluation(BaseModel):
    """Structured per-answer evaluation from Gemini."""
    question_id: str = ""
    technical_correctness: float = Field(default=0.0, ge=0.0, le=1.0)
    depth: float = Field(default=0.0, ge=0.0, le=1.0)
    reasoning: float = Field(default=0.0, ge=0.0, le=1.0)
    specificity: float = Field(default=0.0, ge=0.0, le=1.0)
    communication: float = Field(default=0.0, ge=0.0, le=1.0)
    evidence_corroboration: float = Field(default=0.0, ge=0.0, le=1.0)
    contradiction: float = Field(default=0.0, ge=0.0, le=1.0)
    confidence: float = Field(default=0.0, ge=0.0, le=1.0)
    brief_explanation: str = ""
    follow_up_needed: bool = False
    suggested_follow_up: str = ""
    next_question: str = ""
    next_question_reason: str = ""
    target_competency: str = ""
    question_type: str = ""
    demonstrated: List[str] = []
    missing: List[str] = []
    misconceptions: List[str] = []


class AnswerInterviewQuestionOut(BaseModel):
    id: str
    competency: str
    prompt: str
    follow_ups: List[str] = []


class AnswerInterviewResponse(BaseModel):
    session_id: str
    question_id: str
    next_action: str = "next"
    action: Optional[str] = None
    spoken_response: Optional[str] = None
    ai_available: bool = True
    evaluation: Optional[AnswerInterviewEvaluation] = None
    evaluation_pending: bool = False
    current_index: int = 0
    current_question: Optional[AnswerInterviewQuestionOut] = None
    completed: bool = False
    answered_count: int = 0
    total_questions: int = 0
    note: Optional[str] = None
    failure_category: Optional[str] = None
    provider_status: Optional[int] = None
    provider_code: Optional[str] = None
    retry_after: Optional[str] = None

File: frontend/src/services/assessment.ts
import { apiFetch } from "./api";

/**
 * INAURA direct skill assessment.
 *
 * Prototype instrument: a short deterministic quiz used as the strongest
 * evidence source (VERY HIGH reliability), not a validated measurement of
 * human proficiency.
 */

export type AssessmentReason =
  | "conflicting_evidence"
  | "priority_gap"
  | "low_confidence"
  | "role_requirement"
  | string;

export type LastAssessment = {
  attempt_id: string | null;
  score: number;
  correct_count: number;
  question_count: number;
  completed_at: string | null;
  assessment_version: string | null;
  total_attempts: number;
};

export type AssessmentLayerKey = "knowledge" | "practical" | "interview";

export type LayerCapability = {
  skill: string;
  skill_key: string;
  knowledge_assessment: boolean;
  practical_assessment: boolean;
  interview_assessment: boolean;
  practical_kind: string | null;
  layers: AssessmentLayerKey[];
};

export type LayerResult = {
  status: string;
  score: number | null;
  attempt_id?: string | null;
  session_id?: string | null;
  completed_at?: string | null;
  validity?: string | null;
  communication_score?: number | null;
};

export type AvailableAssessment = {
  skill: string;
  skill_key: string;
  reason: AssessmentReason;
  reason_label: string;
  proficiency: number;
  confidence: number;
  required_level: number;
  gap: number;
  importance: number;
  priority_score: number;
  evidence_count: number;
  already_assessed: boolean;
  question_count: number;
  last_assessment: LastAssessment | null;
  /** Applicable assessment layers for this skill (additive; may be absent). */
  capabilities?: LayerCapability | null;
  /** Per-layer completion state (additive; may be absent). */
  layers?: Partial<Record<AssessmentLayerKey, LayerResult>> | null;
};

export type AvailableAssessmentsResponse = {
  assessment_version: string;
  target_role: string | null;
  analysis_id: string | null;
  available: AvailableAssessment[];
  disclaimer: string;
};

export type AssessmentQuestion = {
  id: string;
  skill: string;
  difficulty: string;
  kind: string;
  prompt: string;
  type: "multiple_choice" | "short_answer" | string;
  options: Array<{ id: string; text: string }>;
  version: string;
};

export type StartAssessmentResponse = {
  attempt_id: string;
  skill: string;
  skill_key: string;
  assessment_version: string;
  question_count: number;
  started_at: string;
  expires_at: string;
  questions: AssessmentQuestion[];
  disclaimer: string;
};

export type QuestionFeedback = {
  question_id: string;
  skill: string;
  difficulty: string;
  kind: string;
  correct: boolean;
  answered: boolean;
  explanation: string;
};

export type SubmitAssessmentResponse = {
  attempt_id: string;
  skill: string;
  assessment_version: string;
  score: number;
  correct_count: number;
  question_count: number;
  validity: "valid" | "low_confidence" | "invalid" | string;
  counts_as_evidence: boolean;
  completed_at: string;
  source_reliability: number;
  feedback: QuestionFeedback[];
  disclaimer: string;
  analysis?: {
    analysis_id?: string | null;
    target_role?: string | null;
    readiness_score?: number | null;
    evidence_component?: number | null;
    skill?: {
      skill: string;
      proficiency: number;
      confidence: number;
      gap: number;
      evidence_proficiency?: number;
      assessment_score?: number | null;
      has_assessment?: boolean;
    } | null;
  } | null;
};

export function getAvailableAssessments() {
  return apiFetch<AvailableAssessmentsResponse>("/analysis/assessment/available");
}

export function startAssessment(skill: string, questionCount?: number) {
  return apiFetch<StartAssessmentResponse>("/analysis/assessment/start", {
    method: "POST",
    body: JSON.stringify(questionCount ? { skill, question_count: questionCount } : { skill }),
  });
}

export function submitAssessment(
  attempt_id: string,
  responses: Record<string, string>,
  duration_seconds?: number
) {
  return apiFetch<SubmitAssessmentResponse>("/analysis/assessment/submit", {
    method: "POST",
    body: JSON.stringify({ attempt_id, responses, duration_seconds }),
  });
}

// ---------------------------------------------------------------------------
// Layer 2 — practical (work-sample) assessment
// ---------------------------------------------------------------------------

export type PracticalTask = {
  id: string;
  skill: string;
  task_type: string;
  title: string;
  prompt: string;
  starter_code: string;
  language: string;
  estimated_minutes: number;
  dimensions: Array<{ id: string; label: string }>;
  check_count: number;
  version: string;
};

export type StartPracticalResponse = {
  attempt_id: string;
  skill: string;
  skill_key: string;
  task_version: string;
  started_at: string;
  task: PracticalTask;
  disclaimer: string;
};

export type PracticalCheckResult = {
  check_id: string;
  dimension: string;
  description: string;
  passed: boolean;
};

export type SubmitPracticalResponse = {
  attempt_id: string;
  skill: string;
  task_id: string;
  task_version: string;
  score: number;
  validity: "valid" | "low_confidence" | "invalid" | string;
  counts_as_evidence: boolean;
  completed_at: string;
  source_reliability: number;
  dimensions: Record<string, number>;
  checks: PracticalCheckResult[];
  analysis?: SubmitAssessmentResponse["analysis"];
  disclaimer: string;
};

export function startPracticalAssessment(skill: string) {
  return apiFetch<StartPracticalResponse>("/analysis/assessment/practical/start", {
    method: "POST",
    body: JSON.stringify({ skill }),
  });
}

export function submitPracticalAssessment(
  attempt_id: string,
  code: string,
  duration_seconds?: number
) {
  return apiFetch<SubmitPracticalResponse>("/analysis/assessment/practical/submit", {
    method: "POST",
    body: JSON.stringify({ attempt_id, code, duration_seconds }),
  });
}

// ---------------------------------------------------------------------------
// Layer 3 — AI skill interview (one session validates exactly one skill)
// ---------------------------------------------------------------------------

export type InterviewQuestion = {
  id: string;
  competency: string;
  prompt: string;
  follow_ups: string[];
};

export type StartInterviewResponse = {
  session_id: string;
  skill: string;
  skill_key: string;
  interview_version: string;
  status: string;
  started_at: string;
  plan: {
    skill: string;
    competencies: Array<{ id: string; label: string }>;
    focus_areas: string[];
    questions: InterviewQuestion[];
    estimated_minutes: number;
  };
  evaluated_dimensions: Record<string, string[]>;
  privacy_notice: string;
  disclaimer: string;
};

export type InterviewSession = {
  session_id: string | null;
  skill: string | null;
  status: string;
  plan: StartInterviewResponse["plan"];
  transcript: Array<{
    question_id: string;
    competency: string;
    prompt: string;
    answer: string;
    answered: boolean;
  }>;
  technical_scores: { per_competency: Record<string, number>; overall: number } | null;
  communication_scores: { per_dimension: Record<string, number>; overall: number } | null;
  started_at: string | null;
  completed_at: string | null;
};

export type CompleteInterviewResponse = {
  session_id: string;
  skill: string;
  status: string;
  validity: string;
  counts_as_evidence: boolean;
  completed_at: string | null;
  source_reliability: number | null;
  technical_scores: InterviewSession["technical_scores"];
  communication_scores: InterviewSession["communication_scores"];
  analysis?: SubmitAssessmentResponse["analysis"];
  note?: string | null;
};

export function startSkillInterview(skill: string) {
  return apiFetch<StartInterviewResponse>("/analysis/assessment/interview/start", {
    method: "POST",
    body: JSON.stringify({ skill }),
  });
}

export function getSkillInterview(session_id: string) {
  return apiFetch<InterviewSession>(`/analysis/assessment/interview/${encodeURIComponent(session_id)}`);
}

export function submitInterviewResponses(session_id: string, responses: Record<string, string>) {
  return apiFetch<{ session_id: string; skill: string | null; answered: number; total: number }>(
    "/analysis/assessment/interview/respond",
    {
      method: "POST",
      body: JSON.stringify({ session_id, responses }),
    }
  );
}

export function completeSkillInterview(session_id: string) {
  return apiFetch<CompleteInterviewResponse>("/analysis/assessment/interview/complete", {
    method: "POST",
    body: JSON.stringify({ session_id }),
  });
}
